runtime/src/consensus/tendermint/verifier/predicates.rs

```diff
@@ -0,0 +1,146 @@
+use anyhow::anyhow;
+use io_context::Context;
+
+use crate::{
+    common::namespace::Namespace,
+    consensus::{
+        beacon::EpochTime,
+        roothash::{Header, HeaderType},
+        state::{
+            beacon::ImmutableState as BeaconState, roothash::ImmutableState as RoothashState,
+            ConsensusState,
+        },
+        tendermint::{verifier::Cache, LightBlockMeta},
+        verifier::Error,
+        LightBlock,
+    },
+};
+
+/// Verifies that the namespace in the runtime header matches the trusted namespace.
+pub fn verify_namespace(trusted: Namespace, runtime_header: &Header) -> Result<(), Error> {
+    if trusted != runtime_header.namespace {
+        return Err(Error::VerificationFailed(anyhow!(
+            "header namespace does not match trusted runtime id"
+        )));
+    }
+
+    Ok(())
+}
+
+/// Verifies that consensus height has correctly advanced since the last update.
+pub fn verify_consensus_advance(cache: &Cache, consensus_block: &LightBlock) -> Result<(), Error> {
+    if consensus_block.height < cache.last_verified_height {
+        // Reject requests for earlier heights.
+        return Err(Error::VerificationFailed(anyhow!(
+            "height seems to have moved backwards"
+        )));
+    }
+
+    Ok(())
+}
+
+/// Verifies that the round has correctly advanced since the last update.
+pub fn verify_round_advance(
+    cache: &Cache,
+    runtime_header: &Header,
+    consensus_block: &LightBlock,
+    epoch: EpochTime,
+) -> Result<(), Error> {
+    if runtime_header.round < cache.last_verified_round {
+        // Reject requests for earlier rounds.
+        return Err(Error::VerificationFailed(anyhow!(
+            "round seems to have moved backwards"
+        )));
+    }
+    if epoch < cache.last_verified_epoch {
+        // Reject requests for earlier epochs.
+        return Err(Error::VerificationFailed(anyhow!(
+            "epoch seems to have moved backwards"
+        )));
+    }
+
+    // If round has advanced make sure that consensus height has also advanced as a round can
+    // only be finalized in a subsequent consensus block. This is to avoid a situation where
+    // one would keep feeding the same consensus block for subsequent rounds.
+    //
+    // NOTE: This needs to happen before the call to verify_consensus_only which updates the
+    //       cache.last_verified_height field.
```

Comment:
Note could be removed or moved as `cache.last_verified_height` is not updated in the `verify_consensus_only` anymore.